### PR TITLE
feat: upstream TSyntax helper functions

### PR DESCRIPTION
**Pull request by kim-em**
From Std.Lean.Syntax.

### CODE REVIEW

File: src/Lean/Syntax.lean [hunk 1]
```diff
@@ -530,4 +530,27 @@ def Stack.matches (stack : Syntax.Stack) (pattern : List $ Option SyntaxNodeKind
...
+
+namespace TSyntax
+
+/--
+Applies the given function to every subsyntax.
+
+Like `Syntax.replaceM` but for typed syntax.
+(Note there are no guarantees of type correctness here.)
+-/
+def replaceM [Monad M] (f : Syntax → M (Option Syntax)) (stx : TSyntax k) : M (TSyntax k) :=
```
Comment (david-christiansen) [inline src/Lean/Syntax.lean:542]:
```suggestion
def replaceM [Monad m] (f : Syntax → M (Option Syntax)) (stx : TSyntax k) : m (TSyntax k) :=
```

Comment (semorrison) [reply src/Lean/Syntax.lean:542]:
```suggestion
def replaceM [Monad m] (f : Syntax → m (Option Syntax)) (stx : TSyntax k) : m (TSyntax k) :=
```
Missed one!

File: src/Lean/Syntax.lean [hunk 2]
```diff
@@ -530,4 +530,27 @@ def Stack.matches (stack : Syntax.Stack) (pattern : List $ Option SyntaxNodeKind
...
 
 end Syntax
+
+namespace TSyntax
+
+/--
+Applies the given function to every subsyntax.
+
+Like `Syntax.replaceM` but for typed syntax.
+(Note there are no guarantees of type correctness here.)
```
Comment (Kha) [inline src/Lean/Syntax.lean:540]: I think I'd prefer the abstraction to just not exist then, unsafely creating `TSyntax` should be explicit

File: src/Lean/Syntax.lean [hunk 3]
```diff
@@ -530,4 +530,27 @@ def Stack.matches (stack : Syntax.Stack) (pattern : List $ Option SyntaxNodeKind
...
+
+end TSyntax
+
+/--
+Constructs a typed separated array from elements.
+The given array does not include the separators.
+
+Like `Syntax.SepArray.ofElems` but for typed syntax.
+-/
+def Syntax.TSepArray.ofElems {sep} (elems : Array (TSyntax k)) : TSepArray k sep :=
```
Comment (Kha) [inline src/Lean/Syntax.lean:553]: We should move this in front of https://github.com/leanprover/lean4/blob/5cc0c3f145ffa623a617ca31387d7f589802c1d3/src/Init/Meta.lean#L566-L567 and make use of it there

**Review by Kha — CHANGES_REQUESTED** [review]

File: src/Lean/Syntax.lean (hunk 3 above)
Comment (semorrison) [reply src/Lean/Syntax.lean:553]: Done.

File: src/Lean/Syntax.lean (hunk 2 above)
Comment (semorrison) [reply src/Lean/Syntax.lean:540]: Turns out it is not even used. I will remove from this PR and maybe clean it up in Std later.

**Pull request merged**